flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/BootstrapTransformation.java

```diff
@@ -0,0 +1,129 @@
+import org.apache.flink.state.api.output.operators.BroadcastStateBootstrapOperator;
+import org.apache.flink.state.api.output.partitioner.HashSelector;
+import org.apache.flink.state.api.output.partitioner.KeyGroupRangePartitioner;
+import org.apache.flink.state.api.runtime.BoundedStreamConfig;
+import org.apache.flink.state.api.runtime.OperatorIDGenerator;
+import org.apache.flink.state.api.runtime.metadata.SavepointMetadata;
+import org.apache.flink.streaming.api.graph.StreamConfig;
+import org.apache.flink.streaming.api.operators.StreamOperator;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/**
+ * Bootstrapped data that can be written into a {@code Savepoint}.
+ * @param <T> The input type of the transformation.
+ */
+@PublicEvolving
+@SuppressWarnings("WeakerAccess")
+public class BootstrapTransformation<T> {
+	private final DataSet<T> dataSet;
+
+	private final SavepointWriterOperatorFactory factory;
+
+	@Nullable
+	private final HashSelector<T> keySelector;
+
+	@Nullable
+	private final TypeInformation<?> keyType;
+
+	BootstrapTransformation(
+		DataSet<T> dataSet,
+		SavepointWriterOperatorFactory factory) {
+		this.dataSet = dataSet;
+		this.factory = factory;
+		this.keySelector = null;
+		this.keyType = null;
+	}
+
+	<K> BootstrapTransformation(
+		DataSet<T> dataSet,
+		SavepointWriterOperatorFactory factory,
+		@Nonnull KeySelector<T, K> keySelector,
+		@Nonnull TypeInformation<K> keyType) {
+		this.dataSet = dataSet;
+		this.factory = factory;
+		this.keySelector = new HashSelector<>(keySelector);
+		this.keyType = keyType;
+	}
+
+	DataSet<TaggedOperatorSubtaskState> getOperatorSubtaskStates(
+		String uid,
+		StateBackend stateBackend,
+		SavepointMetadata metadata,
+		Path savepointPath) {
+		DataSet<T> input = dataSet;
+		if (keySelector != null) {
+			input = dataSet.partitionCustom(new KeyGroupRangePartitioner(metadata.maxParallelism()), keySelector);
+		}
+
+		final StreamConfig config;
+		if (keyType == null) {
+			config = new BoundedStreamConfig();
+		} else {
+			TypeSerializer<?> keySerializer = keyType.createSerializer(dataSet.getExecutionEnvironment().getConfig());
+			config = new BoundedStreamConfig(keySerializer, keySelector);
+		}
+
+		StreamOperator<TaggedOperatorSubtaskState> operator = factory.getOperator(
+			System.currentTimeMillis(),
+			savepointPath);
+
+		operator = dataSet.clean(operator);
+		config.setStreamOperator(operator);
+
+		config.setOperatorName(uid);
+		config.setOperatorID(OperatorIDGenerator.fromUid(uid));
+		config.setStateBackend(stateBackend);
+
+		BoundedOneInputStreamTaskRunner<T> operatorRunner = new BoundedOneInputStreamTaskRunner<>(
+			config,
+			metadata
+		);
+
+		MapPartitionOperator<T, TaggedOperatorSubtaskState> subtaskStates = input
+			.mapPartition(operatorRunner)
+			.name(uid);
+
+		if (operator instanceof BroadcastStateBootstrapOperator) {
+			subtaskStates = subtaskStates.setParallelism(1);
```

Comment:
Because the DataSet api does not support broadcast partitioning. This is the closest approximation; if the parallelism is one than all elements will be shipped to the same task slot. In practice it's fine because we only need a single copy of the broadcast state which can then be restored at any parallelism.